common/functions/src/scalars/maths/abs.rs

```diff
@@ -47,87 +47,69 @@ impl AbsFunction {
 }
 
 macro_rules! impl_abs_function {
-    ($column:expr, $type:ident, $cast_type:expr) => {{
-        let mut series = $column.column().to_minimal_array()?;
-
-        // coerce String to Float
-        if *$column.data_type() == DataType::String {
-            series = series.cast_with_type(&DataType::Float64)?;
-        }
-
-        let primitive_array = series.$type()?;
-        let column: DataColumn = primitive_array
-            .apply_cast_numeric(|v| v.abs())
-            .cast_with_type(&$cast_type)?
-            .into();
-        Ok(column.resize_constant($column.column().len()))
+    ($column:expr, $type:ident) => {{
+        let unary = ScalarUnaryExpression::<$type, $type, _>::new($type::abs);
+        let col = unary.eval($column.column())?;
+        Ok(col.arc())
     }};
 }
 
-impl Function for AbsFunction {
+impl Function2 for AbsFunction {
     fn name(&self) -> &str {
         "abs"
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        let nullable = args.iter().any(|arg| arg.is_nullable());
-        let data_type = if args[0].is_numeric() || args[0].is_string() || args[0].is_null() {
-            Ok(match args[0].data_type() {
-                DataType::Int8 => DataType::UInt8,
-                DataType::Int16 => DataType::UInt16,
-                DataType::Int32 => DataType::UInt32,
-                DataType::Int64 => DataType::UInt64,
-                DataType::String => DataType::Float64,
-                dt => dt.clone(),
-            })
-        } else {
-            Err(ErrorCode::IllegalDataType(format!(
-                "Expected numeric types, but got {}",
-                args[0]
-            )))
-        }?;
-
-        Ok(DataTypeAndNullable::create(&data_type, nullable))
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        assert_numeric(args[0])?;
+        let data_type = match args[0].data_type_id() {
+            TypeID::Int8 => i8::to_data_type(),
+            TypeID::Int16 => i16::to_data_type(),
+            TypeID::Int32 => i32::to_data_type(),
+            TypeID::Int64 => i64::to_data_type(),
+            TypeID::UInt8 => u8::to_data_type(),
+            TypeID::UInt16 => u16::to_data_type(),
+            TypeID::UInt32 => u32::to_data_type(),
+            TypeID::UInt64 => u64::to_data_type(),
+            TypeID::Float32 => f32::to_data_type(),
+            TypeID::Float64 => f64::to_data_type(),
+            _ => unreachable!(),
+        };
```

Comment:
Unlike before, the results are no longer cast here. I think it is reasonable behavior for the result type to be the same as the parameter type.